### ISSUE TITLE
Emdros trips up

### ISSUE THREAD

**Issue by dirkroorda** [emdros/emdros#1]
Hi Ulrik,

In a query like this Emdros raises an exception that can not be caught.

```
select all objects where
[word FOCUS voc_lex_utf8 has 'וָֹ']
```

http://shebanq.ancient-data.org/hebrew/query?id=6336

Shebanq fires this query through the Python3 interface here:

https://gitlab.huc.knaw.nl/tt/shebanq/-/blob/master/src/shebanq/modules/mql.py?ref_type=heads#L110

and I had to write code around it so that the query gets intercepted before reaching the emdros function.

Is this a know issue? Any comment would be appreciated.

Oh, it is in Emdros 3.7.3

Best, Dirk

**Comment by emg** [comment]
Hi Dirk,

sorry for the delay. And thanks for writing.

I ran this through mql(1):

```
ERROR: Compile error while executing query:
--------------------------------------------
select all objects where
[word FOCUS voc_lex_utf8 has 'וָֹ']
 GO
--------------------------------------------
Symbolchecking failed with compiler error
Feature voc_lex_utf8 does not exist on its enclosing object type.
```

Advice:

1) It does not throw an exception. Instead, it reports a symbol checking error. You would detect this by checking both the database error boolean and the compiler error boolean in the two-tuple returned by EmdrosEnv's executeString(). This should work:

```
(bNoDatabaseError, bNoCompilerError) = env.executeString(sanitize(query, msgs), compilerResult, False, False)

if not bNoCompilerError:
    # A compiler error occurred
    error_message = env.getCompilerError()
    # blah blah ...
elif not bNoDatabaseError:
    # A database error occurred
    error_message = env.getCompilerError()
    # blah blah ...
else:
    # No compiler error occurred, and no database error occurred
    # blah blah ...
```
    

2) The database you use may have a different schema from mine. I do not have the feature word.voc_lex_utf8. Instead, I have word.g_voc_lex_utf8. So if I change this to:

```
$ cat tmp.mql
select all objects where
[word FOCUS g_voc_lex_utf8 has 'וָֹ']
```
it throws this exception:

```
$ mql -b 3 -d ~/db/BHS__bhs4_ubs.sqlite3 tmp.mql 
EmdrosException:EMdF/emdf_value.cpp:250:Unknown operator
Program aborted.
```

This is because HAS does not work on string features, only on list features. 

Instead, you probably want to use the regex operator:

```
$ cat tmp.mql
select all objects where
[word FOCUS g_voc_lex_utf8 ~ 'וָֹ']
$ mql -b 3 -d ~/db/BHS__bhs4_ubs.sqlite3 tmp.mql 
 // 

```
3. You may want to catch the EmdrosException in your Python code, which should be wrapped in the Python SWIG wrapper. If you can't get this to work, please let me know.

I will look into catching this in a cleaner way.

**Comment by emg** [comment]
Hi Dirk,

thanks again for reporting this.

I have implemented typechecking safeguards against using HAS or IN when comparing a string with a string in a topographic query.

Implemented as of @c8bc557 on master. Will release in 3.9.1 or whatever the next version after 3.9.0 will be.

Closed.

**Comment by dirkroorda** [comment]
Many thanks for your explanation, Ulrik. This helps me to rectify my code.
